Introduces the new PIF builder for SwiftBuild

pmattos opened the pull request:
Introduces the new PIF builder for packages in `SwiftBuildSupport`.

### Motivation:

This continues the work started by [PR 8271](https://github.com/swiftlang/swift-package-manager/pull/8271), introducing now our new *PIF builder* for packages. As expected, this will happen in the new `SwiftBuildSupport` target, introduced to support the just released **Swift Build** project. 

### Modifications:

This 1st PR *only* includes the new code in the corresponding `SwiftBuildSupport/PIFPackageBuilder*.swift` source files. The code isn't actually being used (or unit tested) just yet.

The code still uses the older `SWBProjectModel.PIF` API to build the PIF data model. In an upcoming PR, I will be migrating this over to the newer `SwiftBuild.ProjectModel` API, and then finally replacing the original code to actually use the PIF builder to send package build information to Swift Build.

PS. I did run `SwiftFormat` in all new code, as indicated by the contributors guide.

### Result:

This marks the start of our *package PIF builder* unification (i.e., between Xcode and SwiftPM).

Tracked by rdar://147767772.

jakepetroules commented:
@swift-ci test

dschaefer2 commented:
Make sure you add the new files to the CMakeLists.txt file.

pmattos commented:
@swift-ci test

jakepetroules commented:
@swift-ci test

pmattos commented:
@swift-ci test

jakepetroules commented:
> /home/build-user/swiftpm/Sources/SwiftBuildSupport/PIFBuilder.swift:1558:52: error: value of type 'ProductType' has no member 'targetType'

What's going on here?

pmattos commented:
@swift-ci test

pmattos commented:
> > /home/build-user/swiftpm/Sources/SwiftBuildSupport/PIFBuilder.swift:1558:52: error: value of type 'ProductType' has no member 'targetType'
> 
> What's going on here?

We had this duplicated `extension ProductType` but since I added `#if canImport(SwiftBuild)` my version vanished, hence the error. I just added back the original extension (for now at least).

jakepetroules commented:
@swift-ci test linux self hosted

pmattos commented:
@swift-ci test macos smoke test

jakepetroules commented:
@swift-ci test windows
